src/pynxtools_xps/reader.py

```diff
@@ -72,439 +72,575 @@
-                    scan_key = detector_key.replace(
-                        "raw_data/raw", f"raw_data/{data_var}"
-                    )
-
-                    template[scan_key] = xr_data[data_var].data
-                    if units:
-                        template[f"{scan_key}/@units"] = units
-
-        # Add multi-dimensional `raw` array for each detector
-        for detector_nm, value in detector_scans.items():
-            raw_key = modified_key.replace(
-                "/DETECTOR[detector]/", f"/DETECTOR[{detector_nm}]/"
-            )
-            template[raw_key] = np.array(value)
+            raise ValueError(XPSReader.__file_err_msg__)
 
+        # If there are multiple input data files of the same type,
+        # make sure that existing keys are not overwritten.
+        existing = [
+            (key, self.xps_data[key], data_dict[key])
+            for key in set(self.xps_data).intersection(data_dict)
+        ]
 
-def fill_template_with_value(key, value, template):
-    """
-    Fill NeXus template with a key-value pair.
+        self.xps_data = {**self.xps_data, **data_dict}
+        for key, value1, value2 in existing:
+            self.xps_data[key] = concatenate_values(value1, value2)
 
-    Parameters
-    ----------
-    key : str
-        DESCRIPTION.
-    value :
-        Any value coming from the XPS, config, or ELN file.
-    template : Template
-        A NeXus template.
+        return {}
 
-    """
-    if value is None or str(value) == "None":
-        return
+    def get_entry_names(self) -> List[str]:
+        """
+        Returns a list of entry names which should be constructed from the data.
+        Defaults to creating a single entry named "entry".
+        """
+        # Track entries for using for eln data
+        entries: List[str] = []
 
-    atom_types: List = []
-    if "chemical_formula" in key:
-        atom_types = list(extract_atom_types(value))
+        try:
+            for entry, entry_values in self.xps_data["data"].items():
+                entries += [entry]
+        except KeyError:
+            pass
 
-    if isinstance(value, datetime.datetime):
-        value = value.isoformat()
+        if not entries:
+            entries += ["entry"]
+
+        return list(dict.fromkeys(entries))
+
+    def setup_template(self) -> Dict[str, Any]:
+        """
+        Setups the initial data in the template.
+        This may be used to set fixed information, e.g., about the reader.
+        """
+        return {}
+
+    def handle_objects(self, objects: Tuple[Any]) -> Dict[str, Any]:
+        """
+        Handles the objects passed into the reader.
+        """
+        return {}
+
+    def post_process(self) -> None:
+        """
+        Do postprocessing after all files and the config file are read .
+        """
+        # TODO: make processing of multiple entities robust
+        pass  # self.process_multiple_entities()
```

Comment:
You don't necessarily need those here. But also fine if you want to have them for completeness / future extensions.

Reply:
Yeah, future extensions was the idea 👍 I had something for `setup_template`, but can't find it now. I'll add it later.